src/charm.py

```diff
@@ -699,39 +702,112 @@ def _activate_charm(self) -> None:
-                working_dir=DISCOURSE_PATH,
-                user=CONTAINER_APP_USERNAME,
-                environment=self._create_discourse_environment_settings(),
-                timeout=60,
+        if not container.can_connect():
+            event.fail("Unable to connect to container, container is not ready")
+            return
+
+        email = event.params["email"]
+
+        user_exists = container.exec(
+            [os.path.join(DISCOURSE_PATH, "bin/bundle"), "exec", "rake", "users:exists", email],
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+        )
+        try:
+            user_exists.wait_output()
+        except ExecError:
+            event.fail(f"User with email {email} does not exist")
+            return
+
+        process = container.exec(
+            [
+                os.path.join(DISCOURSE_PATH, "bin/bundle"),
+                "exec",
+                "rake",
+                "admin:create",
+            ],
+            stdin=f"{email}\nn\nY\n",
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+            timeout=60,
+        )
+        try:
+            process.wait_output()
+            event.set_results({"user": email})
+        except ExecError as ex:
+            event.fail(
+                f"Failed to make user with email {email} an admin: {ex.stdout}"  # type: ignore
             )
-            try:
-                process.wait_output()
-                event.set_results({"user": f"{email}"})
-            except ExecError as ex:
-                event.fail(
-                    # Parameter validation errors are printed to stdout
-                    f"Failed to create user with email {email}: {ex.stdout}"  # type: ignore
-                )
-        else:
-            event.fail("Container is not ready")
+
+    def _on_create_user_action(self, event: ActionEvent):
+        """Create a new user in Discourse.
+
+        Args:
+            event: Event triggering the create_user action.
+        """
+        container = self.unit.get_container(CONTAINER_NAME)
+        if not container.can_connect():
+            event.fail("Unable to connect to container, container is not ready")
+            return
+
+        email = event.params["email"]
+        password = self._generate_password(16)
+
+        user_exists = container.exec(
+            [os.path.join(DISCOURSE_PATH, "bin/bundle"), "exec", "rake", "users:exists", email],
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+        )
+        try:
+            user_exists.wait_output()
+            event.fail(f"User with email {email} already exists")
+        except ExecError:
+            pass
```

Comment:
I believe we should differentiate between failure and user does not exist.

Reply:
:+1:  on what @javierdelapuente is saying, and also - I believe also that you might be getting an ExecError with the email existing and that `pass` is effectively masking the `exit 1` you have, so that might be the reason your integration test is failing...

Reply:
:+1:  on what @javierdelapuente is saying, and also - I believe also that you might be getting an ExecError with the email existing and that `pass` is effectively masking the `exit 1` you have, so that might be the reason your integration test is failing...

Reply:
have addressed this, by returning a different error code